src/symtab2gb/symtab2gb_main.cpp

```diff
@@ -0,0 +1,113 @@
+#include <goto-programs/goto_model.h>
+#include <goto-programs/link_goto_model.h>
+#include <goto-programs/write_goto_binary.h>
+#include <iostream>
+#include <json-symtab-language/json_symtab_language.h>
+#include <string>
+#include <util/exception_utils.h>
+#include <util/invariant.h>
+#include <util/optional.h>
+
+static inline bool failed(bool error_indicator)
+{
+  return error_indicator;
+}
+
+static void run_symtab2gb(
+  const std::vector<std::string> &symtab_filenames,
+  const std::string &gb_filename)
+{
+  // try opening all the files first to make sure we can
+  // even read/write what we want
+  auto out_file = std::ofstream{gb_filename};
+  if(!out_file.is_open())
+  {
+    throw system_exceptiont{"couldn't open output file `" + gb_filename + "'"};
+  }
+  std::vector<std::ifstream> symtab_files;
+  for(auto const &symtab_filename : symtab_filenames)
+  {
+    auto symtab_file = std::ifstream{symtab_filename};
+    if(!symtab_file.is_open())
+    {
+      throw system_exceptiont{"couldn't open input file `" + symtab_filename +
+                              "'"};
+    }
+    symtab_files.emplace_back(std::move(symtab_file));
+  }
+
+  auto message_handler = stream_message_handlert{std::cerr};
+
+  auto const symtab_language = new_json_symtab_language();
+  symtab_language->set_message_handler(message_handler);
+
+  auto linked_goto_model = goto_modelt{};
+
+  for(std::size_t ix = 0; ix < symtab_files.size(); ++ix)
+  {
+    auto const &symtab_filename = symtab_filenames[ix];
+    auto &symtab_file = symtab_files[ix];
+    if(failed(symtab_language->parse(symtab_file, symtab_filename)))
+    {
+      throw invalid_source_file_exceptiont{
+        "failed to parse symbol table from file `" + symtab_filename + "'"};
+    }
+    auto symtab = symbol_tablet{};
+    if(failed(symtab_language->typecheck(symtab, "<unused>")))
+    {
+      throw invalid_source_file_exceptiont{
+        "failed to typecheck symbol table from file `" + symtab_filename + "'"};
+    }
+    auto goto_model = goto_modelt{};
+    goto_model.symbol_table = symtab;
+    goto_convert(goto_model, message_handler);
+    link_goto_model(linked_goto_model, goto_model, message_handler);
+  }
+  if(failed(write_goto_binary(out_file, linked_goto_model)))
+  {
+    throw system_exceptiont{"failed to write goto binary to " + gb_filename};
+  }
+}
+
+int main(int argc, char *argv[])
+{
+  PRECONDITION(argc > 0);
+  auto const executable_name = argv[0];
+  try
+  {
+    if(argc < 2)
+    {
+      throw invalid_command_line_argument_exceptiont{
+        "expect input file as first argument", "<positional>"};
+    }
+    std::vector<std::string> symtab_filenames;
+    optionalt<std::string> gb_file;
+    for(int i = 1; i < argc; ++i)
+    {
+      if(argv[i] == std::string("--out"))
+      {
+        if(i + 1 >= argc)
+        {
+          throw invalid_command_line_argument_exceptiont{
+            "expected output file name", "--out"};
```

Comment:
To be honest, this is a bit too simplistic to my taste: I think this should have support for `--help` and `--version` at bare minimum. Therefore I'd recommend inheriting from `parse_options_baset`.